File: webserver/src/main/java/com/llh/webserver/service/impl/SysUserServiceImpl.java
package com.llh.webserver.service.impl;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.util.StrUtil;
import com.llh.webserver.common.exception.BusinessException;
import com.llh.webserver.common.exception.msg.BasicExpEnum;
import com.llh.webserver.model.QSysUser;
import com.llh.webserver.model.SysUser;
import com.llh.webserver.pojo.SimplePageQueryVO;
import com.llh.webserver.repository.SysUserRepo;
import com.llh.webserver.service.SysUserService;
import com.querydsl.core.types.Predicate;
import com.querydsl.core.types.dsl.BooleanExpression;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

/**
 * SysUserServiceImpl
 * <p>
 * CreatedAt: 2020-04-18 17:00
 *
 * @author llh
 */
@Service("sysUserService")
@Slf4j
public class SysUserServiceImpl implements SysUserService {

    @Autowired
    private SysUserRepo userRepo;
    @Autowired
    private PasswordEncoder passwordEncoder;

    @Override
    @Transactional
    public SysUser save(SysUser entity) {
        entity.setCreatedAt(getNow());
        entity.setUpdatedAt(getNow());
        entity.setPassword(passwordEncoder.encode(entity.getPassword()));
        entity.setDataStatus(PERSISTENCE);
        entity.setAccountStatus(ACTIVE_ACCOUNT);
        return userRepo.save(entity);
    }

    @Override
    @Transactional
    public SysUser remove(String id) {
        SysUser user = findById(id);
        user.setDataStatus(REMOVE);
        return userRepo.save(user);
    }

    @Override
    @Transactional
    public SysUser updateById(SysUser entity) {
        SysUser user = findById(entity.getId());
        checkVersion(user, entity);
        if (StrUtil.isNotBlank(entity.getNewPassword()))
            user.setPassword(passwordEncoder.encode(entity.getNewPassword()));
        BeanUtil.copyProperties(entity, user, getCopyOptions());
        user.setUpdatedAt(getNow());
        return userRepo.save(user);
    }

    @Override
    public SysUser findById(String id) {
        Optional<SysUser> optional = userRepo.findById(id);
        if (!optional.isPresent()) {
            log.debug(" 找不到 SysUser 数据，id为：{} 。", id);
            throw new BusinessException(BasicExpEnum.DATA_NONENTITY_ERROR);
        }
        return optional.get();
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        Optional<SysUser> optional = userRepo.findTopByUsername(username);
        if (!optional.isPresent()) {
            log.info("找不到用户名为 {} 的数据。", username);
            throw new BusinessException(BasicExpEnum.Auth_ERROR);
        }
        return optional.get();
    }

    @Override
    public Page<SysUser> pageQuery(SimplePageQueryVO<SysUser> queryVO) {
        if (queryVO == null || queryVO.getPageNumber() == null || queryVO.getPageSize() == null) {
            log.error("分页查询参数缺失：{}。", queryVO);
            throw new BusinessException(BasicExpEnum.DATA_VALIDATE_ERROR);
        }
        Predicate predicate = handleQueryParam(queryVO);
        PageRequest pageRequest = handlePageAndOrderParam(queryVO);
        return null == predicate
            ? userRepo.findAll(pageRequest)
            : userRepo.findAll(predicate, pageRequest); // 第一个参数不能为null，真的麻烦。
    }

    private PageRequest handlePageAndOrderParam(SimplePageQueryVO<SysUser> queryVO) {
        Sort sort = Sort.unsorted();
        if (StrUtil.isNotBlank(queryVO.getOrderField()))
            sort = Sort.by(StrUtil.isNotBlank(queryVO.getOrderType())
                ? Sort.Direction.fromString(queryVO.getOrderType())
                : Sort.Direction.ASC, queryVO.getOrderField());
        return PageRequest.of(queryVO.getPageNumber(), queryVO.getPageSize(), sort);
    }

    private Predicate handleQueryParam(SimplePageQueryVO<SysUser> queryVO) {
        if (null == queryVO.getModel()) return null;
        SysUser model = queryVO.getModel();
        QSysUser queryUser = QSysUser.sysUser;
        Predicate predicate = null;
        if (StrUtil.isNotBlank(model.getUsername())) {
            predicate = queryUser.username.like(model.getUsername());
        }
        return predicate;
    }
}

File: server/src/main/kotlin/com/llh/server/service/impl/SysUserServiceImpl.kt
package com.llh.server.service.impl

import com.llh.server.dao.SysUsers
import com.llh.server.model.SysUser
import com.llh.server.model.copyProperties
import com.llh.server.pojo.AccountVO
import com.llh.server.pojo.createEmptyAccount
import com.llh.server.service.SysUserService
import me.liuwj.ktorm.database.Database
import me.liuwj.ktorm.dsl.and
import me.liuwj.ktorm.dsl.eq
import me.liuwj.ktorm.dsl.update
import me.liuwj.ktorm.entity.add
import me.liuwj.ktorm.entity.find
import me.liuwj.ktorm.entity.sequenceOf
import org.apache.logging.log4j.kotlin.Logging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service

/**
 * SysUserServiceImpl
 *
 * CreatedAt: 2020-06-01 23:30
 *
 * @author llh
 */
@Service("sysUserService")
class SysUserServiceImpl : SysUserService, Logging {

    @Autowired
    private lateinit var database: Database


    override fun save(entity: SysUser): SysUser {
        entity.createdAt = getNow()
        entity.updatedAt = getNow()
        entity.accountStatus = activation
        entity.dataStatus = persistence
        database.sequenceOf(SysUsers).add(entity)
        return entity
    }

    override fun remove(id: String): Boolean {
        val updated = database.update(SysUsers) {
            it.dataStatus to remove
            it.updatedAt to getNow()
            where {
                it.id eq id
            }
        }
        return updated > 0
    }

    override fun updateById(entity: SysUser): Boolean {
        val model = findById(entity.id) ?: return false
        model.copyProperties(entity)
        model.updatedAt = getNow()
        return model.flushChanges() > 0
    }

    override fun findById(id: String): SysUser? {
        val find = database.sequenceOf(SysUsers).find {
            it.id eq id and (it.dataStatus eq persistence)
        }
        if (find == null)
            logger.warn("not find user(id:${id}) info")
        return find
    }

    override fun loadUserByUsername(username: String?): UserDetails {
        username ?: return createEmptyAccount()
        return convertAccount(findTopByUsername(username))
    }

    override fun findTopByUsername(username: String): SysUser? {
        return database.sequenceOf(SysUsers)
            .find {
                it.username eq username and (it.dataStatus eq true)
            }
    }


    /**
     * 将用户类转换为帐户类。
     * 默认返回空的帐户类。
     */
    private fun convertAccount(user: SysUser?): AccountVO {
        user ?: return createEmptyAccount()
        return AccountVO(user.username, user.password, user.id)
    }
}
